Fix release build with Qt 5.10

schnitzeltony opened the pull request:
Moving Qt from 5.9.4 -> 5.10.1 causes build failures for mixxx:

| <command-line>:0:20: error: call to non-constexpr function 'void qt_noop()'
| /usr/include/qt5/QtCore/qstringview.h:291:14: note: in expansion of macro 'Q_ASSERT'
|     { return Q_ASSERT(int(size()) == size()), int(size()); }
| .. many-more errors of same kind

This is caused by a different usage of Q_ASSERT in qtbase. For further details
see qtbase

commit 8ea27bb1c669e21100a6a042b0378b3346bdf671
Author: Marc Mutz <marc.mutz@kdab.com>
Date:   Tue Mar 14 22:45:05 2017 +0100

    Make Q_ASSERT() usable in constexpr functions

Fix mixxx by copying/aliging the new definition of Q_ASSERT from qglobal.h
shipped with 5.10.1.

Tested for Qt 5.9.4 and 5.10.1

Signed-off-by: Andreas Müller <schnitzeltony@gmail.com>

schnitzeltony committed:
Fix release build with Qt 5.10

Moving Qt from 5.9.4 -> 5.10.1 causes build failures for mixxx:

| <command-line>:0:20: error: call to non-constexpr function 'void qt_noop()'
| /usr/include/qt5/QtCore/qstringview.h:291:14: note: in expansion of macro 'Q_ASSERT'
|     { return Q_ASSERT(int(size()) == size()), int(size()); }
| .. many-more errors of same kind

This is caused by a different usage of Q_ASSERT in qtbase. For further details
see qtbase

commit 8ea27bb1c669e21100a6a042b0378b3346bdf671
Author: Marc Mutz <marc.mutz@kdab.com>
Date:   Tue Mar 14 22:45:05 2017 +0100

    Make Q_ASSERT() usable in constexpr functions

Fix mixxx by copying/aliging the new definition of Q_ASSERT from qglobal.h
shipped with 5.10.1.

Tested on Linux with Qt 5.9.4 and 5.10.1

Signed-off-by: Andreas Müller <schnitzeltony@gmail.com>

Be-ing commented on build/depends.py:
```diff
@@ -1260,7 +1260,7 @@ def configure(self, build, conf):
             # that we include. We can't define QT_NO_DEBUG because that would
             # mean turning off QDebug output. qt_noop() is what Qt defines
             # Q_ASSERT to be when QT_NO_DEBUG is defined.
```
If we're not using `qt_noop()` anymore, the comment should be edited.

schnitzeltony replied:
I can do that but what about failed autotests?

Be-ing replied:
@sblaisot do you know why AppVeyor is failing?

Be-ing commented:
Building with `qt5=1` still works for me with Qt 5.9.4 on Fedora 27. Could anyone test on macOS? Has anyone gotten Qt5 builds working on Windows?

Be-ing commented:
Before we can merge your code into Mixxx, please sign the [contributor agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ). In a nutshell, this gives us permission to distribute your code under the GPL as well as in the Mac App Store which has some terms incompatible with the GPL.

daschuer commented:
He is already on our contributor list.
@schnitzeltony thank you for continuing improving Mixxx!

daschuer commented:
Unfortunately the library checks are failing now on Windows. 
I cannot remember to see such issue. Could this be a side effect from this branch?  

```
Checking for C library ebur128... )' was unexpected at this time.
no
Checking for C library libebur128... )' was unexpected at this time.
no
Checking for C header file sys/queue.h... )' was unexpected at this time.
no
Configuring PortAudio
Checking for C library portaudio... )' was unexpected at this time.
no
ERROR:root:Unmet dependency: Did not find libportaudio.a, portaudio.lib, or the PortAudio-v19 development header files.
Configuring PortMIDI
Checking for C library porttime... )' was unexpected at this time.
no
Checking for C library libporttime... )' was unexpected at this time.
no
Checking for C header file porttime.h... )' was unexpected at this time.
no
ERROR:root:Unmet dependency: Did not find PortTime or its development headers.
Configuring Qt
Configuring TestHeaders
```

daschuer commented:
)' in the error message could be a sign that 
'Q_ASSERT(x)=static_cast<void>(false&&(x))'
can't be set as a compiler argument. 

Maybe something is treated as an special character. Probably &&.

schnitzeltony commented:
> )' in the error message could be a sign that
> 'Q_ASSERT(x)=static_cast(false&&(x))'
> can't be set as a compiler argument.
> 
> Maybe something is treated as an special character. Probably &&.

Yes I think so and in this condition I would not be happy seeing this patch merged - it definitely breaks Windows build (it was fine before my patch). I will try some other options and if I find something not breaking with Qt5.9 and fixing mixxx with Qt5.10 on Linux I'll send out. Note that I have no Windows PC so I cannot test that path. Might last a bit because the Qt 5.10 machine is at the office and it will not make my boss exactly happy if I spend too much time on mixxx :)

Maybe somebody more skilled (and/or with Windows machine) might want to look at https://github.com/qt/qtbase/commit/8ea27bb1c669e21100a6a042b0378b3346bdf671 that is the commit with a detailed description which I guess fixed the error in qtbase I am seeing with mixxx/Qt5.10.

schnitzeltony commented:
Minor correction: https://github.com/qt/qtbase/commit/8ea27bb1c669e21100a6a042b0378b3346bdf671

1. added further Q_ASSERT in a super enhanced but problematic way (from my limited sight)
2. changed Q_ASSERT to avoid the breakages.

We have 2. now when building with Qt >= 5.10 and the world would be happy if Windows's python  or compiler (?) wouldn't behave differently... sigh

schnitzeltony commented:
FYI:

replace

`static_cast<void>`

by

`(void)`

did not solve Windows issue

schnitzeltony commented:
In further builds I detected another problem with Qr5.10 (have no logs here). It seems not yet a big issue to others so I would like to postpone this due to other challenges. I keep request open as reminder and for those just building with linux and hope to come back later.

Be-ing commented:
We are releasing 2.1 as the last Qt4 release, so there is no rush to complete this for the 2.1 release. So I am targeting this for the 2.2 milestone.

daschuer commented:
Does 
-D'Q_ASSERT(x)=static_cast<void>(false and (cond))'
work with Visual Studio?

rryan commented:
Fixing this in the 2.1 branch may be important, since Debian unstable is
dropping Qt 4 and we don't want them to also drop Mixxx.

On Tue, May 1, 2018 at 3:31 PM, Daniel Schürmann <notifications@github.com>
wrote:

> Does
> -D'Q_ASSERT(x)=static_cast(false and (cond))'
> work with Visual Studio?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1526#issuecomment-385810000>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABnn5GsRTiqYoQBf4YrEVXJ7LGZOtAmks5tuOIwgaJpZM4STqbi>
> .
>

Be-ing commented:
Is this still relevant?

rryan commented:
I fixed it in master with https://github.com/mixxxdj/mixxx/commit/ba6d352ac9d3d073f5f9fa662bc26aa3aed49b0e but we could also fix it in 2.1. If a distro is choosing between dropping 2.1.x or shipping it with Qt 5 then I guess I'd prefer they shipped it with Qt 5, so fixing the build in the 2.1 branch is reasonable.

rryan commented:
Oh this is targeted at master. For some reason I thought it was targeting the 2.1 branch.

rryan commented:
e.g. https://bugs.launchpad.net/mixxx/+bug/1768148

Be-ing commented:
Should we retarget this at 2.1?

daschuer commented:
IMHO yes. 
I am only in doubt that this branch will work.

Be-ing commented:
Obsoleted by #1632 and #1684.
